pulsar/consumer_partition.go

```diff
@@ -875,6 +868,20 @@ func (pc *partitionConsumer) discardCorruptedMessage(msgID *pb.MessageIdData,
 		})
 }
 
+func (pc *partitionConsumer) getCompressionProvider(compressionType pb.CompressionType) (
+	provider compression.Provider, ok bool) {
+	if pc.compressionProviders == nil {
+		pc.compressionProviders = map[pb.CompressionType]compression.Provider{
```

Comment:
Will each consumer each have their own reference to all compressions providers? If so why not just preallocate when the partition consumer is created?

Reply:
Not all providers are thread safe. For example currently the zstdProvider reusing its encoder when calling `Compress(data)`. Therefore I think each consumer and each producer should have their own set of compression providers to avoid racing on the global shared providers. And yes, it will have overhead.

Yes, preallocating will make the code here simpler, I will change that.

Reply:
Do you see a race condition? Looking at the source code it seems like it is thread safe?
https://github.com/klauspost/compress/blob/master/zstd/encoder.go#L415

Am I missing something?

Reply:
Oh, Sorry. It is my mistake that doesn't check on the source. The EncodeAll is thread safe.
I think I can close this PR and issue now.